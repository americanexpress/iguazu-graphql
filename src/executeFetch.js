/*
 * Copyright 2018 American Express
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import merge from 'deepmerge';

import config from './config';
import * as types from './duck/types';
import addTypeNamesToQuery from './gql/addTypeNamesToQuery';

function findTypenames(obj, typenames = []) {
  if (!obj || typeof obj !== 'object') {
    return typenames;
  }

  if (Array.isArray(obj)) {
    obj.forEach(field => findTypenames(field, typenames));
  } else {
    Object.entries(obj).forEach(([key, val]) => {
      if (key === '__typename') {
        typenames.push(val);
      } else {
        findTypenames(val, typenames);
      }
    });
  }

  return typenames;
}

async function extractDataFromResponse(res) {
  const body = await res.json();
  const { status } = res;

  if (!res.ok) {
    return Promise.reject(Object.assign(new Error(`${res.statusText} (${res.url})`), { body, status }));
  }

  const { data, errors } = body;
  const typenames = findTypenames(data).filter((v, i, a) => i === a.indexOf(v));

  return Promise.resolve({ data, errors, typenames });
}

async function getAsyncData({
  endpointName, getState, query, variables,
}) {
  const { defaultOpts = {} } = config;
  const { fetch, opts: endpointOpts = {} } = config.getEndpoint(endpointName);
  const { url, opts: fetchOpts = {} } = fetch(getState());
  const { baseFetch } = config;
  const queryWithTypeNames = addTypeNamesToQuery(query);

  const mergedOpts = merge.all([
    {
      method: 'POST',
      headers: {
        'content-type': 'application/json',
        accepts: 'application/json',
      },
    },
    defaultOpts,
    endpointOpts,
    fetchOpts,
    {
      body: JSON.stringify({
        query: queryWithTypeNames,
        variables,
      }),
    },
  ]);

  const res = await baseFetch(url, mergedOpts);
  return extractDataFromResponse(res);
}

function waitAndDispatchFinished(promise, action) {
  return async (dispatch) => {
    try {
      const { data, errors, typenames } = await promise;
      dispatch(Object.assign({}, action, {
        type: types[`${action.type}_FINISHED`], data, error: errors, typenames,
      }));
    } catch (error) {
      dispatch(Object.assign({}, action, { type: types[`${action.type}_ERROR`], error }));
    }
  };
}

export default function executeFetch({
  actionType, endpointName, query, variables,
}) {
  return (dispatch, getState) => {
    const promise = getAsyncData({
      endpointName, getState, query, variables,
    });
    dispatch({
      type: types[`${actionType}_STARTED`], endpointName, query, variables, promise,
    });
    dispatch(waitAndDispatchFinished(promise, {
      type: actionType, endpointName, query, variables,
    }));

    return promise;
  };
}
