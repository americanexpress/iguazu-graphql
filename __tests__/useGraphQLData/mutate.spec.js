/*
 * Copyright 2018 American Express
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import executeFetch from '../../src/executeFetch';

import mutateGraphQLData from '../../src/useGraphQLData/mutate';

jest.mock('../../src/executeFetch', () => jest.fn(() => Promise.resolve('executing fetch')));

describe('mutateGraphQLData', () => {
  const mutation = `
    mutation incrementTheCounter($by: Int){
      incrementCounter(by: $by) {
        count
        offer {
          canRefuse
        }
      }
    }
  `;

  it('is a function', () => expect(mutateGraphQLData).toBeInstanceOf(Function));

  it('requires an endpointName', () => {
    expect(() => mutateGraphQLData({})).toThrowErrorMatchingSnapshot();
  });

  it('requires a mutation', () => {
    expect(() => mutateGraphQLData({
      endpointName: 'sample-endpoint',
    })).toThrowErrorMatchingSnapshot();
  });

  it('requires variables to be an object', () => {
    expect(() => mutateGraphQLData({
      endpointName: 'sample-endpoint',
      mutation,
      variables: true,
    })).toThrowErrorMatchingSnapshot();
  });

  it('returns a function', () => {
    expect(mutateGraphQLData({
      endpointName: 'sample-endpoint',
      mutation,
    })).toBeInstanceOf(Function);
  });


  it('dispatches a new fetch', () => {
    executeFetch.mockClear();
    const variables = { by: 2 };
    const dispatch = jest.fn(v => v);
    const thunk = mutateGraphQLData({ endpointName: 'endpoint-name', mutation, variables });
    thunk(dispatch);
    expect(executeFetch).toHaveBeenCalledTimes(1);
    expect(executeFetch.mock.calls[0][0]).toEqual({
      actionType: 'MUTATION',
      endpointName: 'endpoint-name',
      query: mutation,
      variables,
    });
  });

  it('dispatches a new fetch defaulting to variables of null', () => {
    executeFetch.mockClear();
    const dispatch = jest.fn(v => v);
    const thunk = mutateGraphQLData({ endpointName: 'endpoint-name', mutation });
    thunk(dispatch);
    expect(executeFetch).toHaveBeenCalledTimes(1);
    expect(executeFetch.mock.calls[0][0]).toEqual({
      actionType: 'MUTATION',
      endpointName: 'endpoint-name',
      query: mutation,
      variables: null,
    });
  });

  it('dispatches the thunk from executeFetch', () => {
    executeFetch.mockClear();
    const variables = { by: 2 };
    const dispatch = jest.fn(v => v);
    const thunk = mutateGraphQLData({ endpointName: 'endpoint-name', mutation, variables });
    thunk(dispatch);
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch.mock.calls[0][0]).toBeInstanceOf(Promise);
  });

  it('returns the iguazu interface data', () => {
    executeFetch.mockClear();
    const variables = { by: 2 };
    const dispatch = jest.fn(v => v);
    const thunk = mutateGraphQLData({ endpointName: 'endpoint-name', mutation, variables });
    const iguazuData = thunk(dispatch);
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(iguazuData).toEqual({
      promise: dispatch.mock.calls[0][0],
      status: 'loading',
    });
  });
});
