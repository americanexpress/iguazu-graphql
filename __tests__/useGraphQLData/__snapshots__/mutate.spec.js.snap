// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutateGraphQLData requires a mutation 1`] = `"mutateGraphQLData requires the mutation argument"`;

exports[`mutateGraphQLData requires an endpointName 1`] = `"mutateGraphQLData requires the endpointName argument"`;

exports[`mutateGraphQLData requires variables to be an object 1`] = `"mutateGraphQLData requires the variables argument to be an object"`;
